src/Microsoft.IdentityModel.Protocols.OpenIdConnect/OpenIdConnectProtocolValidator.cs

```diff
@@ -29,22 +29,31 @@
 using System.Collections.Generic;
 using System.ComponentModel;
 using System.Globalization;
+using System.IdentityModel.Tokens.Jwt;
 using System.Linq;
 using System.Security.Cryptography;
 using System.Text;
 using Microsoft.IdentityModel.Json.Linq;
 using Microsoft.IdentityModel.JsonWebTokens;
 using Microsoft.IdentityModel.Logging;
 using Microsoft.IdentityModel.Tokens;
+using JwtRegisteredClaimNames = Microsoft.IdentityModel.JsonWebTokens.JwtRegisteredClaimNames;
 
 namespace Microsoft.IdentityModel.Protocols.OpenIdConnect
 {
     /// <summary>
-    /// Delegate for validating additional claims in 'id_token' 
+    /// Delegate for validating additional claims in 'id_token' (of type <see cref="JsonWebToken"/>).
     /// </summary>
-    /// <param name="idToken"><see cref="SecurityToken"/> to validate</param>
+    /// <param name="idToken"><see cref="JsonWebToken"/> to validate</param>
     /// <param name="context"><see cref="OpenIdConnectProtocolValidationContext"/> used for validation</param>
-    public delegate void IdTokenValidator(SecurityToken idToken, OpenIdConnectProtocolValidationContext context);
+    public delegate void JsonWebTokenValidator(JsonWebToken idToken, OpenIdConnectProtocolValidationContext context);
+
+    /// <summary>
+    /// Delegate for validating additional claims in 'id_token' (of type <see cref="JwtSecurityToken"/>).
+    /// </summary>
+    /// <param name="idToken"><see cref="JwtSecurityToken"/> to validate</param>
+    /// <param name="context"><see cref="OpenIdConnectProtocolValidationContext"/> used for validation</param>
+    public delegate void IdTokenValidator(JwtSecurityToken idToken, OpenIdConnectProtocolValidationContext context);
```

Comment:
Consider marking this one as obsolete too.

Reply:
Good suggestion! I just added this in.

Reply:
@GeoK @mafurman could we have a more generalized calling pattern that receives'json'. The validator inspects the oidc reponse message received against the 'json' which represents a validated idtoken. We assume the json was validated.
I am wondering if the OpenIdConnectProtocolValidator could perform all validation functions including signature.

Reply:
This would simplify the calling pattern for runtimes such as asp.net.

Reply:
@brentschmaltz If I'm understanding this correctly, a method would receive `OpenIDConnectMessage` (authorization response), TokenValidationParameters, and other properties, and it would validate an Id Token from an OIDC message and OpenIdConnectProtocol, in the same call. 

I agree that the calling pattern would be simplified, but I'm not sure if this is a right place for such a change as the class name is OpenIdConnect**Protocol**Validator.

We will probably have to account for TokenValidatedEvent which ASP.NET has, that is invoked after IdToken validation and before OIDC protocol validation.

Reply:
@GeoK about the TokenValidedEvent, we may want to think about when it is called. It the token has arrived in an invalid protocol message, should we consider it a valid id_token?

This is an issue we should bring up with asp.net, users may be making decisions on this token being valid.

Let's back off validation for a moment. 
Could we validate the protocol with just json?

Reply:
If you look at issue https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/1131
It is coded up:
```
OnTokenValidated = context =>
{
    var claims = context.SecurityToken.Claims;
    return Task.CompletedTask;
}
```

Reply:
@brentschmaltz So we would add a string (json) representation of an id token, instead of a JsonWebToken, to OpenIdConnectProtocolValidationContext class?
That design would be more future-proof. The drawback is that we would have to parse the json string which introduces some overhead, compared to our current design. I'm not saying that we shouldn't do it, just trying to point out pros/cons.

Reply:
> 
> 
> If you look at issue #1131
> It is coded up:
> 
> ```
> OnTokenValidated = context =>
> {
>     var claims = context.SecurityToken.Claims;
>     return Task.CompletedTask;
> }
> ```

I don't think that this is a production code. My guess is that a client was debugging and trying to figure out if it could get the claims from an access token.

Reply:
@GeoK We could use a string or define an interface IJsonWebToken. That would be OK as OIDC requires a JWT.

Reply:
@GeoK My thoughts about OnTokenValidated are centered on how / when should we report the jwt in the OIDC response as valid? A jwt arriving over 'Beraer' as an access_token can be valid by checking iss, aud, signature etc. But a jwt arriving as part of a OIDC response should trusted be used if the 'nonce' is not valid.

Are we sending the wrong message here.

Reply:
@brentschmaltz  - This event represents an extensibility point that allows customers to put some custom logic after the id token was validated. Only in a case when a customer explicitly populates context.Result (HandleRequestResult) in their custom logic, OIDC protocol will **_not_** be validated.
Otherwise, `ValidateAuthenticationResponse` method will be called.

Perhaps OnTokenValidated event should be invoked _after_  OIDC protocol validation.

I'm wondering if there are users who use this event to prepare/set the nonce to `OpenIdConnectProtocolValidationContext`. They would be broken if OnTokenValidated event is not invoked before OIDC protocol validation.